zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/schema/AbstractSchema.scala

```diff
@@ -184,20 +185,51 @@ trait Measurement {
 
 //TODO when UI get the schema, it should know which is dimension/measure, functions that can apply onto it, etc.
 // so that it won't ask for a inapplicable function such that get the max to a string field
+abstract class AbstractSchema(typeName: String,
+                              dimension: Seq[Field],
+                              measurement: Seq[Field],
+                              primaryKey: Seq[Field]
+                             ) {
+
+  private val dimensionMap: Map[String, Field] = dimension.map(f => f.name -> f).toMap
+  private val measurementMap: Map[String, Field] = measurement.map(f => f.name -> f).toMap
+
+  val fieldMap: Map[String, Field] = dimensionMap ++ measurementMap ++ Map(AllField.name -> AllField)
+
+  def getTypeName: String = typeName
+  def getTimeField: Option[TimeField]
+  def copySchema: AbstractSchema
+  def toUnresolved: UnresolvedSchema
+}
+
 case class Schema(typeName: String,
                   dimension: Seq[Field],
                   measurement: Seq[Field],
                   primaryKey: Seq[Field],
                   timeField: TimeField
-                 ) {
+                 ) extends AbstractSchema(typeName, dimension, measurement, primaryKey){
 
-  private val dimensionMap: Map[String, Field] = dimension.map(f => f.name -> f).toMap
-  private val measurementMap: Map[String, Field] = measurement.map(f => f.name -> f).toMap
+  override def getTimeField: Option[TimeField] = Some(timeField)
```

Comment:
I suppose `isInstanceOf[Schema]` is actually the same as `hasTimeField`, which we removed previously?